test/unit/data_file_test.rb

```diff
@@ -489,4 +489,25 @@ class DataFileTest < ActiveSupport::TestCase
     assert data_file.respond_to?(:data_format_annotations)
     assert data_file.respond_to?(:data_type_annotations)
   end
+
+  test 'related samples include extracted samples and samples linked in attributes' do
+    project = FactoryBot.create(:project)
+    df1 = FactoryBot.create(:data_file)   # No samples
+    df2 = FactoryBot.create(:data_file)   # Extracted samples
+    df3 = FactoryBot.create(:data_file)   # Linked in sample's attributes only
+    df4 = FactoryBot.create(:data_file)   # Extracted samples and linked in sample's attributes
```

Comment:
use decriptive variable names instead of code comments

Reply:
Resolved in [d93f6df](https://github.com/seek4science/seek/commit/d93f6dff068a259b4b0f59ce619ea5790d7a4645)